src/main/java/com/epam/aidial/core/function/CollectResponseAttachmentsFn.java

```diff
@@ -0,0 +1,117 @@
+package com.epam.aidial.core.function;
+
+import com.epam.aidial.core.Proxy;
+import com.epam.aidial.core.ProxyContext;
+import com.epam.aidial.core.config.ApiKeyData;
+import com.epam.aidial.core.storage.BlobStorageUtil;
+import com.epam.aidial.core.util.ProxyUtil;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import io.vertx.core.Future;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+
+import java.net.URI;
+import java.util.HashSet;
+import java.util.Set;
+
+import static com.epam.aidial.core.util.ProxyUtil.collectAttachedFile;
+
+@Slf4j
+public class CollectResponseAttachmentsFn extends BaseResponseFunction {
+    public CollectResponseAttachmentsFn(Proxy proxy, ProxyContext context) {
+        super(proxy, context);
+    }
+
+    @Override
+    public Future<Void> apply(ObjectNode tree) {
+        try {
+            Set<String> attachments = collectAttachments(tree);
+            if (attachments.isEmpty()) {
+                return Future.succeededFuture();
+            }
+            String perRequestKey = context.getApiKeyData().getPerRequestKey();
+            return proxy.getApiKeyStore().updatePerRequestApiKey(perRequestKey, json -> updateApiKeyData(json, attachments, perRequestKey));
+        } catch (Throwable e) {
+            return Future.failedFuture(e);
+        }
+    }
+
+    private Set<String> collectAttachments(ObjectNode tree) {
```

Comment:
can we move this logic to the utility and cover with tests?

Reply:
fixed